### PR TITLE
feat: Add flagz endpoint for kube-proxy

### LINKED ISSUE #128984: Add flagz endpoint for kube-proxy
### What would you like to be added?

Part of 
- https://github.com/kubernetes/enhancements/issues/4828

Add the `/flagz` endpoint for kube-proxy

Sample response:
```
----------------------------
title: Kubernetes Flagz
description: Command line flags that Kubernetes component was started with.
----------------------------

default-watch-cache-size=100
delete-collection-workers=1
enable-garbage-collector=true
encryption-provider-config-automatic-reload=false
...
```

/sig instrumentation

### Why is this needed?

Refer to [Motivation setcion](https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/4828-component-flagz/README.md#motivation) in KEP.

### PR DESCRIPTION
**Pull request by yongruilin**
#### What type of PR is this?

/kind feature

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/128984 and https://github.com/kubernetes/enhancements/issues/4828

#### Does this PR introduce a user-facing change?

```release-note
Added a `/flagz` endpoint for kube-proxy
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
Example response:
```
curl http://localhost:10249/flagz
```
```
kube-proxy flags
Warning: This endpoint is not meant to be machine parseable, has no formatting compatibility guarantees and is for debugging purposes only.

bind-address=0.0.0.0
bind-address-hard-fail=false
boot-id-file=/proc/sys/kernel/random/boot_id
cleanup=false
cluster-cidr=
config=/var/lib/kube-proxy/config.conf
config-sync-period=15m0s
conntrack-max-per-core=32768
conntrack-min=131072
conntrack-tcp-be-liberal=false
conntrack-tcp-timeout-close-wait=1h0m0s
conntrack-tcp-timeout-established=24h0m0s
conntrack-udp-timeout=0s
conntrack-udp-timeout-stream=0s
```

```docs
- [KEP] : https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/4828-component-flagz/README.md
```

### CODE REVIEW

**Comment by aojea** [comment]
For context @danwinship I talked offline with the author and in kube-proxy we found three options:
- use healthz bind address, but I'm afraid this used for lb healthchecks and we have this dual vs single work that I suggested not to use
- use metrics bind address, the metrics endpoint seems a more stable endpoint to get information of the component
- bind in a new address, we already have enough problems with two addresses :)

/assign @danwinship 
LGTM but waiting for Dan opinion

**Comment by Henrywu573** [comment]
cc @Henrywu573

**Comment by danwinship** [comment]
Is there not a standard for this in other components?
ISTM that it should use the metrics IP, because the information is similarly low-level / debug-ish / potentially-sensitive, as compared to the healthz info.

**Comment by yongruilin** [comment]
@danwinship Thanks for the comment. There are PRs created for other components. I'm also using metrics' IP there.

**Comment by sftim** [comment]
/release-note-edit

```release-note
Added a `/flagz` endpoint for kube-proxy
```

File: cmd/kube-proxy/app/server.go [hunk 1]
```diff
@@ -258,6 +261,12 @@ func newProxyServer(ctx context.Context, config *kubeproxyconfig.KubeProxyConfig
 		return nil, err
 	}
 
+	if utilfeature.DefaultFeatureGate.Enabled(zpagesfeatures.ComponentFlagz) {
+		if flagzReader != nil {
+			s.flagz = flagzReader
+		}
+	}
```
Comment (danwinship) [inline cmd/kube-proxy/app/server.go:268]: You don't need to check the feature gate here; `flagzReader` can only have been set if the feature gate is enabled.

(So you don't even need this as a separate section; you can just unconditionally set `flags: flagzReader` in the initial creation of the `ProxyServer` at the top of the function)

File: cmd/kube-proxy/app/server.go [hunk 2]
```diff
@@ -459,6 +468,12 @@ func serveMetrics(ctx context.Context, bindAddress string, proxyMode kubeproxyco
 
 	configz.InstallHandler(proxyMux)
 
+	if utilfeature.DefaultFeatureGate.Enabled(zpagesfeatures.ComponentFlagz) {
+		if flagzReader != nil {
+			flagz.Install(proxyMux, "kube-proxy", flagzReader)
```
Comment (danwinship) [inline cmd/kube-proxy/app/server.go:473]: likewise, you don't need to check the feature gate here (though in this case you do still need the `nil` check)

**Review by danwinship — CHANGES_REQUESTED** [review]

File: cmd/kube-proxy/app/server.go (hunk 1 above)
Comment (yongruilin) [reply cmd/kube-proxy/app/server.go:265]: Updated. thanks!

File: cmd/kube-proxy/app/server.go (hunk 2 above)
Comment (yongruilin) [reply cmd/kube-proxy/app/server.go]: Updated.

**Comment by danwinship** [comment]
/lgtm
/approve
thanks!

**Pull request merged**